Add toggle to disable automatic seeking to object after placement

MaxKruse opened the pull request:
fixes #22668 

Test for mania specifically included in `TestScenePlacementBeforeTrackStart`, unsure if another place would be better for this.

MaxKruse committed:
Add setting to disable automatic seeking after object placement

MaxKruse committed:
Add test for seeking setting in mania placement test

bdach commented on osu.Game/Screens/Edit/Editor.cs:
```diff
@@ -185,6 +185,7 @@ protected override IReadOnlyDependencyContainer CreateChildDependencies(IReadOnl
 
         private Bindable<float> editorBackgroundDim;
         private Bindable<bool> editorHitMarkers;
+        private Bindable<bool> editorSeekToHitobject;
```
nit: rename to `editorSeekToHitObject` (goes everywhere else too, I'll spare us both posting the same comment everywhere)

bdach commented on osu.Game/Rulesets/Edit/HitObjectComposer.cs:
```diff
@@ -70,6 +71,7 @@ public abstract partial class HitObjectComposer<TObject> : HitObjectComposer, IP
         private FillFlowContainer togglesCollection;
 
         private IBindable<bool> hasTiming;
+        protected Bindable<bool> SeekToHitobject { get; private set; }
```
why `protected`? looks like `private` would suffice

bdach commented on osu.Game/Rulesets/Edit/HitObjectComposer.cs:
```diff
@@ -365,6 +369,9 @@ public void EndPlacement(HitObject hitObject, bool commit)
             {
                 EditorBeatmap.Add(hitObject);
 
+                // conditionally seek based on setting
```
remove comment - it restates the code and brings no value

bdach commented on osu.Game/Localisation/EditorStrings.cs:
```diff
@@ -19,6 +19,11 @@ public static class EditorStrings
         /// </summary>
         public static LocalisableString ShowHitMarkers => new TranslatableString(getKey(@"show_hit_markers"), @"Show hit markers");
 
+        /// <summary>
+        /// "Seek to Object after placement"
+        /// </summary>
+        public static LocalisableString SeekToHitobject => new TranslatableString(getKey(@"seek_to_hitobject"), @"Seek to Object after placement");
```
random capitalisation

```suggestion
        /// <summary>
        /// "Seek to object after placement"
        /// </summary>
        public static LocalisableString SeekToHitObject => new TranslatableString(getKey(@"seek_to_hit_object"), @"Seek to object after placement");
```

\+ update references to match

MaxKruse committed:
fix formating in SeekOnNote test

MaxKruse committed:
change bindable seekToHitObject to private

MaxKruse committed:
naming change from Hitobject to HitObject

bdach committed:
Combine config and time checks into one

Functionally equivalent right now, but the combined variant is more
localised to what it actually needs to do, and less error-prone if
any new code gets appended to the method.

bdach committed:
Rename test scene to match contents

It does not only test "placement before track start" anymore.

bdach committed:
Rewrite test cases

- Depend less on arbitrary timings
- Remove unnecessary seeks
- Change method name to make more sense
- Use nunit style assertions

bdach commented on osu.Game/Screens/Edit/Editor.cs:
```diff
@@ -329,6 +331,10 @@ private void load(OsuConfigManager config)
                                             {
                                                 State = { BindTarget = editorHitMarkers },
+                                            },
+                                            new ToggleMenuItem(EditorStrings.SeekToHitObject)
```
This is in the "view" menu, but it isn't really a "view" option, so it kind of doesn't belong. Not sure what to do about this. @ppy/team-client your thoughts would be appreciated.

bdach reviewed (commented):
This needed more cleanup. I've cleaned up as much as I knew how to clean up, but I am going to need to get one UX thing past a second pair of eyes.

peppy commented:
@bdach is this ready for review?

bdach commented:
@peppy Yeah I was in the middle of getting through this when you typed the comment :sweat_smile: I think it should be relatively alright except for the one open question from the review above.

bdach committed:
Rename moved test method to describe its purpose better

peppy commented:
One thing to note going forward is we will probably add an "auto advance" to osu! editor at some point, ie. one beat forward from actual placement. Which is slightly different to this option but also applied for the same reason.

So one way we could handle this is renaming the option to "Seek after placement" to generalise for that case to. In other words, remove the part which implies this will always seek *to the object specifically*.

bdach commented:
Sure, but to me that's still not really a "view" option, because it doesn't affect the editor visually like any other option in that menu, it changes its behaviour.

Could just repurpose the "view" menu into "options" and call it a day? Dunno. I don't much care, it can live in "view" for now, I just wanted to raise it as a point I find weird in how things are written right now.

peppy commented:
I think under "view" is fine for now until we have more menus. I wouldn't be wanting to add new top level menus just for one item yet.

bdach commented:
Ok, well, if we want to future-proof this for the auto-advance case we may want to swap the `bool` type of the setting for an actual enum. Not sure if it needs to happen right now.

peppy commented:
Well I was thinking more just change the text before it starts to get localised. I'd still hope it was a bool, but just reads as "Auto-seek on placement" or similar.

bdach commented:
Fair. Are you going to grab that or shall I?

peppy commented:
All yours, I think you'll get the wording better than I can anyway 😅

bdach committed:
Use more generic wording for future-proofing

bdach commented:
I've attempted something, let me know if it meets your expectations.
